core/shared/src/main/scala/fs2/internal/Algebra.scala

```diff
@@ -282,115 +304,173 @@ private[fs2] object Algebra {
-                      case Some(toClose) =>
-                        closeAndGo(toClose)
-                      case None =>
-                        // scope already closed, continue with current scope
-                        go(scope, f(Right(())))
-                    }
+                  }
                 }
+
+              scope.findSelfOrAncestor(close.scopeId) match {
+                case Some(toClose) => closeAndGo(toClose)
+                case None =>
+                  scope.findSelfOrChild(close.scopeId).flatMap {
+                    case Some(toClose) =>
+                      closeAndGo(toClose)
+                    case None =>
+                      // scope already closed, continue with current scope
+                      def result =
+                        close.interruptedScope
+                          .map { Result.interrupted _ tupled }
+                          .getOrElse(Result.unit)
+                      go(scope, f(result))
+                  }
               }
 
           }
 
       }
+
     }
 
     F.flatMap(go(scope, stream)) {
-      case Done(_)                  => F.pure(None)
-      case Out(head, scope, tail)   => F.pure(Some((head, scope, tail)))
-      case Interrupted(scope, next) => compileLoop(scope, next)
-      case OpenInterruptibly(scope, concurrent, onInterrupt, next) =>
-        F.flatMap(scope.open(Some((concurrent, onInterrupt)))) { childScope =>
-          compileLoop(childScope, next(Right(childScope)))
+      case Done(_)                => F.pure(None)
+      case Out(head, scope, tail) => F.pure(Some((head, scope, tail)))
+      case Interrupted(_, err) =>
+        err match {
+          case None      => F.pure(None)
+          case Some(err) => F.raiseError(err)
         }
     }
   }
 
+  /**
+    * Inject interruption to the tail used in flatMap.
+    * Assures that close of the scope is invoked if at the flatMap tail, otherwise switches evaluation to `interrupted` path
+    *
+    * @param stream             tail to inject interuption into
+    * @param interruptedScope   scopeId to interrupt
+    * @param interruptedError   Addiitional finalizer errors
+    * @tparam F
+    * @tparam O
+    * @return
+    */
+  def interruptBoundary[F[_], O](
+      stream: FreeC[Algebra[F, O, ?], Unit],
+      interruptedScope: Token,
+      interruptedError: Option[Throwable]
+  ): FreeC[Algebra[F, O, ?], Unit] =
+    stream.viewL.get match {
+      case _: FreeC.Result.Pure[Algebra[F, O, ?], Unit] =>
+        FreeC.interrupted(interruptedScope, interruptedError)
+      case failed: FreeC.Result.Fail[Algebra[F, O, ?], Unit] =>
+        Algebra.raiseError(
+          CompositeFailure
+            .fromList(interruptedError.toList :+ failed.error)
+            .getOrElse(failed.error))
+      case interrupted: FreeC.Result.Interrupted[Algebra[F, O, ?], _, Unit] =>
+        // impossible
+        FreeC.interrupted(interrupted.context, interrupted.deferredError)
```

Comment:
Does this mean this case is unreachable? 
If so, wouldn't it be better to `sys.error` like in the other cases?

Reply:
Yes, but here we are able to recover ok. I mean there is no harm when compiler wll reach this state although I believe it is not possible to enter it. 

Hence therefore, we just do not fail here.

Reply:
I know that it's harmless, I just find it a bit confusing (seeing code forces me to think about that branch, even though it's impossible to get there). Anyway I leave this up to you :)